From f303e2cfab893dbe6357dc4897804c1cb35c6ddc Mon Sep 17 00:00:00 2001
From: Pritesh Patel <pritesh.patel@einfochips.com>
Date: Wed, 9 Apr 2025 07:15:16 +0000
Subject: [PATCH 474/652] riscv: module: fix compilation error of kvrealloc

Backporting the patch "riscv: module: Optimize PLT/GOT entry counting" to kernel 6.6.77
leads to compilation error. Error below:

arch/riscv/kernel/module-sections.c: In function 'module_frob_arch_sections':
arch/riscv/kernel/module-sections.c:172:35: error: too few arguments to function 'kvrealloc'
  172 |                         scratch = kvrealloc(scratch, scratch_size, GFP_KERNEL);
      |                                   ^~~~~~~~~

Hence, adding the changes to resolve this compilation error.

Link: https://lore.kernel.org/linux-riscv/20250409024519.454828-1-samuel.holland@sifive.com
Signed-off-by: Pritesh Patel <pritesh.patel@einfochips.com>
---
 arch/riscv/kernel/module-sections.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/riscv/kernel/module-sections.c b/arch/riscv/kernel/module-sections.c
index 91d4f0fbd0af..a191ba7f8892 100644
--- a/arch/riscv/kernel/module-sections.c
+++ b/arch/riscv/kernel/module-sections.c
@@ -168,10 +168,10 @@ int module_frob_arch_sections(Elf_Ehdr *ehdr, Elf_Shdr *sechdrs,
 		 * close together, so sort a copy of the section to avoid interfering.
 		 */
 		if (sechdrs[i].sh_size > scratch_size) {
-			scratch_size = sechdrs[i].sh_size;
-			scratch = kvrealloc(scratch, scratch_size, GFP_KERNEL);
+			scratch = kvrealloc(scratch, scratch_size, sechdrs[i].sh_size, GFP_KERNEL);
 			if (!scratch)
 				return -ENOMEM;
+			scratch_size = sechdrs[i].sh_size;
 		}
 
 		/* sort relocations requiring a PLT or GOT entry so duplicates are adjacent */
-- 
2.54.0

